### PR TITLE
Issue #7673: empty buckets with min_doc_count = 0 and negative post_zone.

### PR DESCRIPTION
**Pull request by glelouarn**
Here is a correction proposal for issue #7673.

### CODE REVIEW

**Comment by glelouarn** [comment]
Hello, some news according my proposal?

Regards,

Gildas.

**Comment by clintongormley** [comment]
CLA checked.  @jpountz please could you review

File: src/main/java/org/elasticsearch/common/rounding/TimeZoneRounding.java [hunk 1]
```diff
@@ -279,7 +279,14 @@ public long valueForKey(long time) {
...
-            return durationField.add(value, 1);
+            long nextRoundingValue = durationField.add(value, 1);
+            long postTzOffsetVal = postTz.getOffset(value);
+            
+            if (postTzOffsetVal < 0) {                
+                return field.roundCeiling(nextRoundingValue) + postTzOffsetVal;
+            } else {
+                return nextRoundingValue;
+            }
         }
```
Comment (jpountz) [inline src/main/java/org/elasticsearch/common/rounding/TimeZoneRounding.java]: Hmm, this looks a bit hacky to me... What about doing something like:

```java
        public long nextRoundingValue(long currentWithPostZone) {
            long currentWithoutPostZone = postTz.convertLocalToUTC(currentWithPostZone, true);
            long nextWithoutPostZone = durationField.add(currentWithoutPostZone, 1);
            long nextWithPostZone = postTz.convertUTCToLocal(nextWithoutPostZone);
            return nextWithPostZone;
    }
```

**Comment by jpountz** [comment]
@glelouarn Thank you for providing a fix. I provided a suggestion hoping it would make the computation slightly more robust. Also as @cbuescher noted on https://github.com/elasticsearch/elasticsearch/issues/9062#issuecomment-72906463 we should probably fix the other rounding impls that make use of time zones?

**Comment by jpountz** [comment]
Also note for self: We should probably avoid using DateTimeZone.getOffset manually and instead always use DateTimeZone.convertLocalToUTC and DateTimeZone.convertUTCToLocal which are more readable and robust.

**Comment by cbuescher** [comment]
Just an additional thought: by the way nextRoundingValue() is used internally (InternalHistogram.addEmptyBuckets()) it is importand that it only returns values that are compatible with the Rounding used for that aggregation. When - for whatever reason - nextRoundingValue() is called with an input that is not produced by that Rounding, the result will also be off (see my comment in #7673 for that). To make sure that this will not happen maybe it would be good to not just add some duration but also check that the result is actually valid in the current rounding scheme.

**Comment by cbuescher** [comment]
@glelouarn @jpountz I will close this PR since the fix for this issue was part of #9790 and is on 1.4 and 1.5 branch.

**Pull request closed without merging**